### PR TITLE
feat: add filter to check invalid chars in user input

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
- add a filter to prevent attacks with malformed character encodings and control characters (null byte)

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Filters/InvalidChars.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+     *
+     * @param array|null $arguments
+     *
+     * @return void
+     */
+    public function before(RequestInterface $request, $arguments = null)
+    {
+        if ($request->isCLI()) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/Filters/InvalidChars.php:46]: Just because `CSRF` has the code:
https://github.com/codeigniter4/CodeIgniter4/blob/4995720062ede1ec1fd2e5e595dee7abb6e7556e/system/Filters/CSRF.php#L46-L48

File: system/Filters/InvalidChars.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            return $value;
+        }
+
+        if (mb_check_encoding($value, 'UTF-8')) {
+            return $value;
+        }
+
+        throw new RuntimeException(
+            'Invalid UTF-8 characters in ' . $this->source . ': ' . $value
+        );
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/Filters/InvalidChars.php]: I added SecurityException static named constructors.

**Review by MGatner — APPROVED** [review]
Thanks for those changes, looks great!

**Comment by kenjis** [comment]
@MGatner I rebased this PR and added 3 commits.
If you're okay, I will merge this.

> Have you benchmarked this or checked Debug Toolbar timelines to see impact?

I saw Debug Toolbar timelines, but small input does not impact.
Before Filters take 0.0x ms.

File: system/Filters/InvalidChars.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+     * @var string
+     */
+    protected $source;
+
+    /**
+     * Regular expressions for valid control codes
+     *
+     * @var string
+     */
+    protected $controlCodeRegex = '/\A[\r\n\t[:^cntrl:]]*\z/u';
```
Comment (paulbalandan) [inline system/Filters/InvalidChars.php:40]: I think this should not be changeable by child classes and should be private?

File: user_guide_src/source/incoming/filters.rst [hunk 4]
```diff
@@ -193,6 +193,6 @@ In this example, the array ``['dual', 'noreturn']`` will be passed in ``$argumen
 Provided Filters
 ****************
 
-Three filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, and ``DebugToolbar``.
+These filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, ``DebugToolbar`` and ``InvalidChars``.
```
Comment (paulbalandan) [inline user_guide_src/source/incoming/filters.rst:196]:
```suggestion
The filters bundled with CodeIgniter4 are: ``Honeypot``, ``CSRF``, ``DebugToolbar``, and ``InvalidChars``.
```

File: tests/system/Filters/InvalidCharsTest.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    public function testCheckControlStringWithLineBreakAndTabReturnsTheString($input)
+    {
+        $_GET['val'] = $input;
+
+        $this->invalidChars->before($this->request);
+    }
+
+    public function stringWithLineBreakAndTabProvider()
+    {
+        return [
```
Comment (paulbalandan) [inline tests/system/Filters/InvalidCharsTest.php:136]:
```suggestion
        yield from [
```

File: tests/system/Filters/InvalidCharsTest.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        $this->expectExceptionMessage('Invalid Control characters in get:');
+
+        $_GET['val'] = $input;
+
+        $this->invalidChars->before($this->request);
+    }
+
+    public function stringWithControlCharsProvider()
+    {
+        return [
```
Comment (paulbalandan) [inline tests/system/Filters/InvalidCharsTest.php:162]:
```suggestion
        yield from [
```

File: user_guide_src/source/incoming/filters.rst [hunk 7]
```diff
@@ -193,6 +193,6 @@ In this example, the array ``['dual', 'noreturn']`` will be passed in ``$argumen
...
 
 .. note:: The filters are executed in the declared order  that is defined in the config file, but there is one exception to this and it concerns the ``DebugToolbar``, which is always executed last. This is because ``DebugToolbar`` should be able to register everything that happens in other filters.
```
Comment (paulbalandan) [inline user_guide_src/source/incoming/filters.rst:198]: Although not part of the PR but still..
```suggestion
.. note:: The filters are executed in the order defined in the config file. However, if enabled, ``DebugToolbar`` is always executed last because it should be able to capture everything that happens in the other filters.
```

File: tests/system/Filters/InvalidCharsTest.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            ["String contains \t and \r line \n break."],
+        ];
+    }
+
+    /**
+     * @dataProvider stringWithControlCharsProvider
+     *
+     * @param string $input
+     */
+    public function testCheckControlStringWithControlCharsCausesException($input)
```
Comment (paulbalandan) [inline tests/system/Filters/InvalidCharsTest.php:150]:
```suggestion
    /**
     * @dataProvider stringWithControlCharsProvider
     */
    public function testCheckControlStringWithControlCharsCausesException(string $input)
```

File: tests/system/Filters/InvalidCharsTest.php [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    /**
+     * @doesNotPerformAssertions
+     *
+     * @dataProvider stringWithLineBreakAndTabProvider
+     *
+     * @param string $input
+     */
+    public function testCheckControlStringWithLineBreakAndTabReturnsTheString($input)
```
Comment (paulbalandan) [inline tests/system/Filters/InvalidCharsTest.php:127]:
```suggestion
     * @dataProvider stringWithLineBreakAndTabProvider
     */
    public function testCheckControlStringWithLineBreakAndTabReturnsTheString(string $input)
```

File: system/Filters/InvalidChars.php (hunk 3 above)
Comment (kenjis) [reply system/Filters/InvalidChars.php:40]: A user may want to change acceptable control codes.
It is by application specification.
So I changed it as protected property.

Comment (paulbalandan) [reply system/Filters/InvalidChars.php:40]: 👍

**Review by paulbalandan — APPROVED** [review]

**Comment by MGatner** [comment]
💪😊👍
